spec/p2p/p2p.qnt

```diff
@@ -0,0 +1,140 @@
+            AddressBook' = AddressBook,
+            peerState' = peerState,
+            Environment' = Environment,
+            IsUnconditional' = IsUnconditional,
+        }
+    }
+
+    action transportAccept(a: Address):bool = all {
+        or {
+            peersInbound < MaxNumInboundPeers,
+            IsUnconditional.contains(a)
+        },
+        peerState' = peerState.put(a, "inbound"),  //shall we check if it is already outbound and open?
+        connectionState' = connectionState.put(a, "open"),
+        AddressBook' = AddressBook.union(Set(a)) ,
+        all { //unchanged
+            Environment' = Environment,
+            IsUnconditional' = IsUnconditional,
+        }
+    }
+
+    // triggered by a reactor
+    action stopPeerForError(a:Address):bool = all {
+        peerState' = Peers.exclude(Set(a)).mapBy(x => peerState.get(x)),
+        connectionState' = Peers.exclude(Set(a)).mapBy(x => connectionState.get(x)),
+        all { //unchanged
+            AddressBook' = AddressBook,
+            Environment' = Environment,
+            IsUnconditional' = IsUnconditional,
+        }
+    }
+        // inform all reactors then
+    
+
+    action ensurePeers(a:Address):bool =
+        all {
+                require(peersOutbound + peersDialing < MaxNumOutboundPeers),
+                require(not(Peers.contains(a))),
+                all {
+                    // I guess we don't remove AddressBook' = AddressBook.exclude(Set(a)),
+                    //dialPeer(addr)
+                    peerState' = peerState.put(a, "dialing"), 
+                    connectionState' = connectionState.put(a, "init"),
+                    all { //unchanged
+                        AddressBook' = AddressBook,
+                        Environment' = Environment,
+                        IsUnconditional' = IsUnconditional,
+                    }
+                }
+            }
+
+    action dialPeers(S:Set[Address]):bool = // check what the code actually is doing. Now we don't override existing peers in this spec
+        all {
+            peerState' = Peers.union(S).mapBy(a => if (Peers.contains(a)) peerState.get(a) else "dialing"), 
+            connectionState' = Peers.union(S).mapBy(a => if (Peers.contains(a)) connectionState.get(a) else "init"),
+            all { //unchanged
+                        AddressBook' = AddressBook,
+                        Environment' = Environment,
+                        IsUnconditional' = IsUnconditional,
+                    }
+        }
+
+    action init =
+        all {
+            peerState' = Map(),
+            connectionState' = Map(),
+            AddressBook' = Set({ID: "igor", ipAddress: "127.0.0.1"}, {ID: "josef", ipAddress: "127.0.0.2"}),
+            Environment' = Set({ID: "lasaro", ipAddress: "127.0.0.7"}, {ID: "daniel", ipAddress: "127.0.0.6"}),
+            IsUnconditional' = Set({ID: "zarko", ipAddress: "127.0.0.42"}),
+        }
+
+    action step = any {
+        all {
```

Comment:
The `ensurePeers` semantics provides it should choose an address at random to dial. Here this step is done outside this method, right?

Reply:
Yes. This is for debugging purposes. Like this I can tell the REPL on the command line which peer to chose. 
I guess it makes sense to have another definition that pulls the choice in for defining the steps of the system...